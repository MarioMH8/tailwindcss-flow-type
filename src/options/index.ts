export { default as DEFAULT_OPTIONS } from './options.default';
export { default as parseOptions } from './options.functions';
export * from './options.type';
