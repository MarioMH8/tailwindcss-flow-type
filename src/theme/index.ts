export { default as createTextTheme } from './create-text-theme';
export { FLOW_TYPE_THEME_LINE_HEIGHT_DEFAULT, FLOW_TYPE_THEME_TEXT_DEFAULT } from './flow-theme.default';
export type { TailwindFlowTypeTheme } from './flow-theme.type';
export { TAILWIND_FLOW_TYPE_SIZES } from './flow-theme.type';
