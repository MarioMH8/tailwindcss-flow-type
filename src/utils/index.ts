export { default as calculateModularScale } from './calculate-modular-scale';
export { default as isNumber } from './is-number';
