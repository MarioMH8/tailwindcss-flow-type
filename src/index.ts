import type { ThemeConfig } from 'tailwindcss/plugin';
import createPlugin from 'tailwindcss/plugin';

import type { TailwindFlowTypeOptions } from './options';
import { parseOptions } from './options';
import { createTextTheme, FLOW_TYPE_THEME_LINE_HEIGHT_DEFAULT, FLOW_TYPE_THEME_TEXT_DEFAULT } from './theme';
import { calculateModularScale, isNumber } from './utils';

export * from './options';
export * from './theme';
export * from './utils';

export default createPlugin.withOptions<Partial<TailwindFlowTypeOptions>>(
	userOptions => {
		const options = parseOptions(userOptions);

		return tailwind => {
			tailwind.matchUtilities(
				{
					[`${options.prefix}-text`]: value => {
						const fontSize =
							typeof value === 'string' ? (isNumber(value) ? Number(value) : value) : undefined;

						if (fontSize === undefined) {
							return {};
						}

						return {
							fontSize:
								typeof fontSize === 'number' ? calculateModularScale(fontSize, options) : fontSize,
						};
					},
				},
				{
					supportsNegativeValues: false,
					values: tailwind.theme('flow-text') as Record<string, string>,
				}
			);
			tailwind.matchUtilities(
				{
					[`${options.prefix}-text`]: value => {
						return {
							lineHeight: value,
						};
					},
				},
				{
					supportsNegativeValues: false,
					values: tailwind.theme('flow-line-height') as Record<string, string>,
				}
			);
		};
	},
	userOptions => {
		const options = parseOptions(userOptions);

		const theme: ThemeConfig = {
			'flow-line-height': FLOW_TYPE_THEME_LINE_HEIGHT_DEFAULT,
			'flow-text': FLOW_TYPE_THEME_TEXT_DEFAULT,
		};

		if (options.override) {
			theme['text'] = createTextTheme(options);
		}

		return {
			theme,
		};
	}
);
