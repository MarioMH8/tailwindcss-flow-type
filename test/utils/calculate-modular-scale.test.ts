import { describe, expect, it } from 'bun:test';

import { calculateModularScale, DEFAULT_OPTIONS } from '../../src';

describe('tailwindcss-flow-type', () => {
	describe('calculateModularScale', () => {
		describe('with default options', () => {
			it.each([-4, -1, 0, 1, 4])('should works as expected with %p scale value', value => {
				const result = calculateModularScale(value);
				expect(result).toMatchSnapshot();
			});
		});

		describe('with custom units', () => {
			it.each([-4, -1, 0, 1, 4])('should works as expected with %p scale value', value => {
				const result = calculateModularScale(value, {
					...DEFAULT_OPTIONS,
					unit: 'px',
				});
				expect(result).toMatchSnapshot();
			});
		});

		describe('with ratioMin and ratioMax are 1', () => {
			it.each([-4, -1, 0, 1, 4])('should works as expected with %p scale value', value => {
				const result = calculateModularScale(value, {
					...DEFAULT_OPTIONS,
					ratioMax: 1,
					ratioMin: 1,
				});
				expect(result).toMatchSnapshot();
			});
		});

		describe('with screenMin equals screenMax', () => {
			it.each([-4, -1, 0, 1, 4])('should works as expected with %p scale value', value => {
				const result = calculateModularScale(value, {
					...DEFAULT_OPTIONS,
					screenMax: 20,
					screenMin: 20,
				});
				expect(result).toMatchSnapshot();
			});
		});

		describe('with fontSizeMin and fontSizeMax are 0', () => {
			it.each([-4, -1, 0, 1, 4])('should works as expected with %p scale value', value => {
				const result = calculateModularScale(value, {
					...DEFAULT_OPTIONS,
					fontSizeMax: 0,
					fontSizeMin: 0,
				});
				expect(result).toMatchSnapshot();
			});
		});
	});
});
