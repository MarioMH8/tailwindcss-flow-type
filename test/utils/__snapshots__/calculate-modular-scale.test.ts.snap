// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`tailwindcss-flow-type calculateModularScale with default options should works as expected with -4 scale value 1`] = `"clamp(0.7023319615912209rem, calc(0.7023319615912209rem + ((0.7023319615912209 - 0.7023319615912209) * ((100vw - 20rem) / (96 - 20)))), 0.7023319615912209rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with default options should works as expected with -1 scale value 1`] = `"clamp(1.0416666666666667rem, calc(1.0416666666666667rem + ((1.0416666666666667 - 1.0416666666666667) * ((100vw - 20rem) / (96 - 20)))), 1.0416666666666667rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with default options should works as expected with 0 scale value 1`] = `"clamp(1.125rem, calc(1.125rem + ((1.25 - 1.125) * ((100vw - 20rem) / (96 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with default options should works as expected with 1 scale value 1`] = `"clamp(1.265625rem, calc(1.265625rem + ((1.5 - 1.265625) * ((100vw - 20rem) / (96 - 20)))), 1.5rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with default options should works as expected with 4 scale value 1`] = `"clamp(1.802032470703125rem, calc(1.802032470703125rem + ((2.5919999999999996 - 1.802032470703125) * ((100vw - 20rem) / (96 - 20)))), 2.5919999999999996rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with custom units should works as expected with -4 scale value 1`] = `"clamp(0.7023319615912209px, calc(0.7023319615912209px + ((0.7023319615912209 - 0.7023319615912209) * ((100vw - 20px) / (96 - 20)))), 0.7023319615912209px)"`;

exports[`tailwindcss-flow-type calculateModularScale with custom units should works as expected with -1 scale value 1`] = `"clamp(1.0416666666666667px, calc(1.0416666666666667px + ((1.0416666666666667 - 1.0416666666666667) * ((100vw - 20px) / (96 - 20)))), 1.0416666666666667px)"`;

exports[`tailwindcss-flow-type calculateModularScale with custom units should works as expected with 0 scale value 1`] = `"clamp(1.125px, calc(1.125px + ((1.25 - 1.125) * ((100vw - 20px) / (96 - 20)))), 1.25px)"`;

exports[`tailwindcss-flow-type calculateModularScale with custom units should works as expected with 1 scale value 1`] = `"clamp(1.265625px, calc(1.265625px + ((1.5 - 1.265625) * ((100vw - 20px) / (96 - 20)))), 1.5px)"`;

exports[`tailwindcss-flow-type calculateModularScale with custom units should works as expected with 4 scale value 1`] = `"clamp(1.802032470703125px, calc(1.802032470703125px + ((2.5919999999999996 - 1.802032470703125) * ((100vw - 20px) / (96 - 20)))), 2.5919999999999996px)"`;

exports[`tailwindcss-flow-type calculateModularScale with ratioMin and ratioMax are 1 should works as expected with -4 scale value 1`] = `"clamp(1.25rem, calc(1.25rem + ((1.25 - 1.25) * ((100vw - 20rem) / (96 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with ratioMin and ratioMax are 1 should works as expected with -1 scale value 1`] = `"clamp(1.25rem, calc(1.25rem + ((1.25 - 1.25) * ((100vw - 20rem) / (96 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with ratioMin and ratioMax are 1 should works as expected with 0 scale value 1`] = `"clamp(1.125rem, calc(1.125rem + ((1.25 - 1.125) * ((100vw - 20rem) / (96 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with ratioMin and ratioMax are 1 should works as expected with 1 scale value 1`] = `"clamp(1.125rem, calc(1.125rem + ((1.25 - 1.125) * ((100vw - 20rem) / (96 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with ratioMin and ratioMax are 1 should works as expected with 4 scale value 1`] = `"clamp(1.125rem, calc(1.125rem + ((1.25 - 1.125) * ((100vw - 20rem) / (96 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with screenMin equals screenMax should works as expected with -4 scale value 1`] = `"clamp(0.7023319615912209rem, calc(0.7023319615912209rem + ((0.7023319615912209 - 0.7023319615912209) * ((100vw - 20rem) / (20 - 20)))), 0.7023319615912209rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with screenMin equals screenMax should works as expected with -1 scale value 1`] = `"clamp(1.0416666666666667rem, calc(1.0416666666666667rem + ((1.0416666666666667 - 1.0416666666666667) * ((100vw - 20rem) / (20 - 20)))), 1.0416666666666667rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with screenMin equals screenMax should works as expected with 0 scale value 1`] = `"clamp(1.125rem, calc(1.125rem + ((1.25 - 1.125) * ((100vw - 20rem) / (20 - 20)))), 1.25rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with screenMin equals screenMax should works as expected with 1 scale value 1`] = `"clamp(1.265625rem, calc(1.265625rem + ((1.5 - 1.265625) * ((100vw - 20rem) / (20 - 20)))), 1.5rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with screenMin equals screenMax should works as expected with 4 scale value 1`] = `"clamp(1.802032470703125rem, calc(1.802032470703125rem + ((2.5919999999999996 - 1.802032470703125) * ((100vw - 20rem) / (20 - 20)))), 2.5919999999999996rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with fontSizeMin and fontSizeMax are 0 should works as expected with -4 scale value 1`] = `"clamp(0rem, calc(0rem + ((0 - 0) * ((100vw - 20rem) / (96 - 20)))), 0rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with fontSizeMin and fontSizeMax are 0 should works as expected with -1 scale value 1`] = `"clamp(0rem, calc(0rem + ((0 - 0) * ((100vw - 20rem) / (96 - 20)))), 0rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with fontSizeMin and fontSizeMax are 0 should works as expected with 0 scale value 1`] = `"clamp(0rem, calc(0rem + ((0 - 0) * ((100vw - 20rem) / (96 - 20)))), 0rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with fontSizeMin and fontSizeMax are 0 should works as expected with 1 scale value 1`] = `"clamp(0rem, calc(0rem + ((0 - 0) * ((100vw - 20rem) / (96 - 20)))), 0rem)"`;

exports[`tailwindcss-flow-type calculateModularScale with fontSizeMin and fontSizeMax are 0 should works as expected with 4 scale value 1`] = `"clamp(0rem, calc(0rem + ((0 - 0) * ((100vw - 20rem) / (96 - 20)))), 0rem)"`;
